lib/pbench/server/api/resources/datasets_inventory.py

```diff
@@ -63,22 +64,13 @@ def _get(self, params: ApiParams, _) -> Response:
         except TarballNotFound as e:
             raise APIAbort(HTTPStatus.NOT_FOUND, str(e))
 
-        if target is None:
-            file_path = tarball.tarball_path
-        else:
-            dataset_location = tarball.unpacked
-            if dataset_location is None:
-                raise APIAbort(HTTPStatus.NOT_FOUND, "The dataset is not unpacked")
-            file_path = dataset_location / target
+        tarball_name = tarball.tarball_path.name
+        target_path = (
+            tarball_name if target is None else (os.path.join(tarball_name, target))
+        )
+        try:
+            file_path = Tarball.extract(str(tarball.tarball_path), target_path)
```

Comment:
`Tarball.extract` doesn't return a "path" ... there *is* no "path" to a file within the tarball. It returns a byte string with the full *contents* of the file extracted from the tarball.

This model actually worked for something small like `metadata.log`, but may not be viable for larger files. However, in any case, the `send_file` helper probably isn't going to work. (You're hiding this in the unit tests with your mocks.)

I suspect we're going to want to open the file as a *stream* and return the stream, which the caller can use to read the contents (e.g., for internal consumption of `metadata.log`) or pass on to `Flask`.

This logic also won't work for the actual tarball, if the API is called without a relative path. The old code passed the `tarball.tarball_path` to `send_file`, which would return it as a stream through HTTP. You're trying to extract that from the tarball, which isn't what we want. Instead, I think that without a relative path you just want to `return send_file(tarball.tarball_path)`; whereas with a path you need to return the *stream*. Note that `tarfile.extractfile()` returns a buffered reader, which I think could be used directly: you'd be deferring the `.read().decode()` to the caller, and it might be handy to refactor this into `Tarball.extractstream` and `Tarball.extractcontents` (or some such names), with the latter calling the former and doing the read/decode like the current method does.

Reply:
> `Tarball.extract` doesn't return a "path" ... there _is_ no "path" to a file within the tarball. It returns a byte string with the full _contents_ of the file extracted from the tarball.

Oh yes, I don't know how I missed that. I think I became really confused weather I am returning a file or file_path.

Reply:
I will try testing it in the actual environment and rework as required.

Reply:
> if the API is called without a relative path

if the API id called without a relative path then `file not found` error will be thrown. We will require a relative path for the target.

Reply:
> I think that without a relative path you just want to `return send_file(tarball.tarball_path)`

if there is no relative path are we unpacking and returning whole tarball?

Reply:
> it might be handy to refactor this into `Tarball.extractstream` and `Tarball.extractcontents`

`Tarball.extractstream` should return file contents and `Tarball.extractcontents` should return whole list of contents inside a directory?

Reply:
> > it might be handy to refactor this into `Tarball.extractstream` and `Tarball.extractcontents`
> 
> `Tarball.extractstream` should return file contents and `Tarball.extractcontents` should return whole list of contents inside a directory?

I threw out names without a whole lot of thought, and maybe that was unnecessarily confusing.

My point is that the current extract method uses the `tarfile` method, which returns a *byte stream*, and then does a `read().decode()` on the stream to produce a string containing the entire contents of the file. That was simple and expedient when what I was really looking for was an encapsulation of pulling `metadata.log` out of the tarball. It doesn't generalize very well for larger files.

First, note that `/api/v1/inventory` only allows access to "regular files" within the tarball: it'll fail with `UNSUPPORTED_MEDIA_TYPE` if pointed at a directory. So there's never an issue with "directory tree" or anything. Without a relative path following the resource ID, we return the source tarball, as-is; with a path to a regular file within the tarball, we return the contents of that file.

However, unlike the current extract method, we want to return a *stream* which the client can process in chunks, not a single large block. Since the `tarfile.extractfile(path)` API already gives us that stream, you apparently should be able to just use that as the first parameter to `send_file` (documentation says it takes a byte stream as well as a file path). But for some other uses, (e.g., the `metadata.log` access) we need to consume the stream anyway, and we know it's not "enormous" so we might as well encapsulate the current extract logic where it's convenient.

So maybe,

```python
    def filestream(self, path: str) -> IO[bytes]:
        try:
            return tarfile.open(self.tarball_path, "r:*").extractfile(path)
        except Exception as exc:
            raise BadFilePath(path) from exc

    def filecontents(self, path: str) -> str:
        return self.filestream(path).read().decode()
```

(Note that the existing code raises a `MetadataError`, which is now inappropriate since it was designed specifically to complain about accessing `metadata.log`; we need something more general here, so I made up a `BadFilePath`.)

Reply:
BTW: from `tarfile.extractfile` documentation:

> Extract a member from the archive as a file object. member may be a filename or a [TarInfo](https://docs.python.org/3/library/tarfile.html#tarfile.TarInfo) object. If member is a regular file or a link, an [io.BufferedReader](https://docs.python.org/3/library/io.html#io.BufferedReader) object is returned. For all other existing members, [None](https://docs.python.org/3/library/constants.html#None) is returned. If member does not appear in the archive, [KeyError](https://docs.python.org/3/library/exceptions.html#KeyError) is raised.

So if you're relying on `tarfile` to diagnose our unsupported media type, that's actually just a `None` return from `.extractfile`, whereas `KeyError` possibly deserves a special `except KeyError` to report a bad file path. (And maybe any other exception *is* just an "unhandled internal server error.)